File: camera_publisher.py
#!/usr/bin/env python3
"""
ROS2 node that consumes camera frames and publishes ROS2 Image topics.
Prefers local GStreamer shared-memory frames (from entrypoint tee branch) to
avoid RTSP H264 re-decode in Python, with RTSP fallback for compatibility.
"""

import cv2
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image, CameraInfo
from cv_bridge import CvBridge
import yaml
import os
import time

class RTSPCameraPublisher(Node):
    def __init__(self):
        super().__init__('rtsp_camera_publisher')
        
        # Parameters from environment or defaults
        self.declare_parameter('rtsp_url', 'rtsp://127.0.0.1:8554/camera')
        self.declare_parameter('shm_socket', '/tmp/ros_frames')
        default_camera_info = os.path.join(os.path.dirname(__file__), 'camera_info.yaml')
        self.declare_parameter('camera_info_file', default_camera_info)
        self.declare_parameter('frame_id', 'camera_optical_frame')
        self.declare_parameter('publish_rate', 30.0)
        self.declare_parameter('width', 1920)
        self.declare_parameter('height', 1080)
        self.declare_parameter('framerate', 60)
        
        self.rtsp_url = self.get_parameter('rtsp_url').value
        self.shm_socket = self.get_parameter('shm_socket').value
        self.camera_info_file = self.get_parameter('camera_info_file').value
        self.frame_id = self.get_parameter('frame_id').value
        publish_rate = self.get_parameter('publish_rate').value
        width = int(self.get_parameter('width').value)
        height = int(self.get_parameter('height').value)
        framerate = int(self.get_parameter('framerate').value)
        
        # Create publishers
        self.image_pub = self.create_publisher(Image, '/camera/image_raw', 1)
        self.camera_info_pub = self.create_publisher(CameraInfo, '/camera/camera_info', 1)
        
        # Primary path: shared memory frames produced by tee branch in entrypoint.sh.
        # This avoids CPU-heavy H264 decode in Python and keeps ROS publishing low-latency.
        shm_pipeline = (
            f"shmsrc socket-path={self.shm_socket} is-live=true do-timestamp=true "
            f"! video/x-raw,format=RGB,width={width},height={height},framerate={framerate}/1 "
            "! videoconvert "
            "! video/x-raw,format=BGR "
            "! queue leaky=downstream max-size-buffers=1 "
            "! appsink drop=1 max-buffers=1 sync=false"
        )

        self.get_logger().info(f"Connecting to shared-memory stream: {self.shm_socket}")
        self.cap = cv2.VideoCapture(shm_pipeline, cv2.CAP_GSTREAMER)
        self.frame_is_bgr = True
        
        # Wait for first frame to ensure connection.
        # If SHM path is not available, fall back to RTSP for compatibility.
        self.connected = False
        for attempt in range(30):
            ret, frame = self.cap.read()
            if ret:
                self.connected = True
                self.get_logger().info("Connected to shared-memory stream")
                break
            self.get_logger().warn(f"Waiting for shared-memory stream (attempt {attempt+1}/30)...")
            time.sleep(1.0)

        if not self.connected:
            self.get_logger().warn("Shared-memory stream unavailable, falling back to RTSP")
            self.cap.release()
            self.cap = cv2.VideoCapture(self.rtsp_url, cv2.CAP_FFMPEG)
            self.cap.set(cv2.CAP_PROP_BUFFERSIZE, 1)
            self.frame_is_bgr = True

            for attempt in range(30):
                ret, frame = self.cap.read()
                if ret:
                    self.connected = True
                    self.get_logger().info("Connected to RTSP fallback stream")
                    break
                self.get_logger().warn(f"Waiting for RTSP fallback (attempt {attempt+1}/30)...")
                time.sleep(1.0)
        
        if not self.connected:
            self.get_logger().error("Failed to connect to both SHM and RTSP streams")
            raise RuntimeError("Camera stream connection failed")
        
        self.bridge = CvBridge()
        
        # Load camera calibration
        self.get_logger().info(f"Using camera_info_file: {self.camera_info_file}")
        self.camera_info = self.load_camera_info()
        self.camera_info.width = width
        self.camera_info.height = height
        
        # Publish at specified rate (default 30 Hz)
        self.create_timer(1.0/publish_rate, self.publish_frame)
        
        self.frame_count = 0
        self.read_fail_count = 0
        self.get_logger().info(f"RTSP Camera Publisher started at {publish_rate} Hz")

    def load_camera_info(self):
        """Load camera calibration from YAML file"""
        try:
            exists = os.path.exists(self.camera_info_file)
            self.get_logger().info(f"Loading camera_info from {self.camera_info_file} (exists={exists})")
            if not exists:
                raise FileNotFoundError(f"Camera info file not found: {self.camera_info_file}")

            with open(self.camera_info_file, 'r') as f:
                config = yaml.safe_load(f) or {}

            # If provided file lacks calibration data, try packaged camera_info.yaml as fallback
            cam_mat = config.get('camera_matrix', {}) or {}
            dist = config.get('distortion_coefficients', {}) or {}
            needs_fallback = not cam_mat.get('data') or not dist.get('data')
            if needs_fallback:
                packaged = os.path.join(os.path.dirname(__file__), 'camera_info.yaml')
                if os.path.exists(packaged):
                    try:
                        with open(packaged, 'r') as pf:
                            pconf = yaml.safe_load(pf) or {}
                        # merge missing keys
                        if not cam_mat.get('data') and pconf.get('camera_matrix'):
                            config['camera_matrix'] = pconf['camera_matrix']
                        if not dist.get('data') and pconf.get('distortion_coefficients'):
                            config['distortion_coefficients'] = pconf['distortion_coefficients']
                        self.get_logger().info(f"Merged missing K/D from packaged camera_info: {packaged}")
                    except Exception:
                        self.get_logger().warn(f"Failed to read packaged camera_info.yaml at {packaged}")

            info = CameraInfo()
            info.header.frame_id = self.frame_id
            info.height = config.get('image_height', 1080)
            info.width = config.get('image_width', 1920)
            info.distortion_model = config.get('distortion_model', 'plumb_bob')

            # Camera matrix K (3x3)
            cam_mat = config.get('camera_matrix', {}) or {}
            info.k = cam_mat.get('data', [1000, 0, 960, 0, 1000, 540, 0, 0, 1])

            # Distortion coefficients D (1x5)
            dist = config.get('distortion_coefficients', {}) or {}
            info.d = dist.get('data', [0, 0, 0, 0, 0])

            self.get_logger().info(f"Loaded camera calibration: {info.width}x{info.height}")
            return info
        except Exception as e:
            self.get_logger().warn(f"Failed to load camera_info from {self.camera_info_file}: {e}")

            # Return default calibration
            info = CameraInfo()
            info.header.frame_id = self.frame_id
            info.height = 1080
            info.width = 1920
            info.distortion_model = 'plumb_bob'
            info.k = [1000, 0, 960, 0, 1000, 540, 0, 0, 1]
            info.d = [0, 0, 0, 0, 0]

            self.get_logger().info("Using default camera calibration")
            return info

    def publish_frame(self):
        """Capture frame from RTSP and publish as ROS2 Image"""
        ret, frame = self.cap.read()
        
        if not ret:
            self.read_fail_count += 1
            if self.read_fail_count % 60 == 0:
                src = "RTSP fallback" if self.frame_is_bgr else "SHM"
                self.get_logger().warn(f"Failed to read frame from {src} ({self.read_fail_count} consecutive fails)")
            return

        self.read_fail_count = 0
        
        msg = self.bridge.cv2_to_imgmsg(frame, encoding='bgr8')
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = self.frame_id
        
        # Update camera info timestamp to match image
        self.camera_info.header.stamp = msg.header.stamp
        
        # Publish both messages
        self.image_pub.publish(msg)
        self.camera_info_pub.publish(self.camera_info)
        
        self.frame_count += 1
        if self.frame_count % 30 == 0:
            self.get_logger().debug(f"Published {self.frame_count} frames")

    def destroy_node(self):
        """Cleanup resources"""
        if self.cap and self.cap.isOpened():
            self.cap.release()
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    node = RTSPCameraPublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: camera_publisher.cpp
#include <chrono>
#include <string>
#include <thread>

#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <sensor_msgs/msg/camera_info.hpp>
#include <camera_info_manager/camera_info_manager.hpp>

#include <gst/gst.h>
#include <gst/app/gstappsink.h>

using namespace std::chrono_literals;

class CameraPublisherNode : public rclcpp::Node
{
public:
  CameraPublisherNode() : Node("rtsp_camera_publisher")
  {
    // Declare parameters
    declare_parameter<std::string>("shm_socket", "/tmp/ros_frames");
    declare_parameter<std::string>("rtsp_url", "rtsp://127.0.0.1:8554/camera");
    declare_parameter<std::string>("camera_info_file", "/etc/camera_info.yaml");
    declare_parameter<std::string>("frame_id", "camera_optical_frame");
    declare_parameter<double>("publish_rate", 30.0);
    declare_parameter<int>("width", 1920);
    declare_parameter<int>("height", 1080);
    declare_parameter<int>("framerate", 30);

    auto shm_socket = get_parameter("shm_socket").as_string();
    rtsp_url_ = get_parameter("rtsp_url").as_string();
    auto camera_info_file = get_parameter("camera_info_file").as_string();
    frame_id_ = get_parameter("frame_id").as_string();
    auto publish_rate = get_parameter("publish_rate").as_double();
    width_ = get_parameter("width").as_int();
    height_ = get_parameter("height").as_int();
    auto framerate = get_parameter("framerate").as_int();

    frame_size_ = width_ * height_ * 3;  // RGB, 3 bytes per pixel

    // Create publishers
    image_pub_ = create_publisher<sensor_msgs::msg::Image>("/camera/image_raw", 1);
    camera_info_pub_ = create_publisher<sensor_msgs::msg::CameraInfo>("/camera/camera_info", 1);

    // Initialize GStreamer
    gst_init(nullptr, nullptr);

    // Try SHM connection first (direct RGB — no videoconvert needed)
    std::string shm_pipeline =
      "shmsrc socket-path=" + shm_socket + " is-live=true do-timestamp=true "
      "! video/x-raw,format=RGB,width=" + std::to_string(width_) +
      ",height=" + std::to_string(height_) +
      ",framerate=" + std::to_string(framerate) + "/1 "
      "! queue leaky=downstream max-size-buffers=1 "
      "! appsink name=sink drop=true max-buffers=1 sync=false emit-signals=false";

    RCLCPP_INFO(get_logger(), "Connecting to shared-memory stream: %s", shm_socket.c_str());

    bool connected = false;
    pipeline_ = gst_parse_launch(shm_pipeline.c_str(), nullptr);
    if (pipeline_) {
      appsink_ = gst_bin_get_by_name(GST_BIN(pipeline_), "sink");
      gst_element_set_state(pipeline_, GST_STATE_PLAYING);

      for (int attempt = 0; attempt < 30; ++attempt) {
        if (try_pull_frame()) {
          connected = true;
          encoding_ = "rgb8";
          RCLCPP_INFO(get_logger(), "Connected to shared-memory stream (rgb8, no videoconvert)");
          break;
        }
        RCLCPP_WARN(get_logger(), "Waiting for shared-memory stream (attempt %d/30)...", attempt + 1);
        std::this_thread::sleep_for(1s);
      }

      if (!connected) {
        cleanup_pipeline();
      }
    }

    // Fallback to RTSP via GStreamer (needs decodebin + videoconvert)
    if (!connected) {
      RCLCPP_WARN(get_logger(), "Shared-memory stream unavailable, falling back to RTSP");

      std::string rtsp_pipeline =
        "rtspsrc location=" + rtsp_url_ + " latency=0 "
        "! rtph264depay ! h264parse ! avdec_h264 "
        "! videoconvert "
        "! video/x-raw,format=RGB "
        "! queue leaky=downstream max-size-buffers=1 "
        "! appsink name=sink drop=true max-buffers=1 sync=false emit-signals=false";

      pipeline_ = gst_parse_launch(rtsp_pipeline.c_str(), nullptr);
      if (pipeline_) {
        appsink_ = gst_bin_get_by_name(GST_BIN(pipeline_), "sink");
        gst_element_set_state(pipeline_, GST_STATE_PLAYING);

        for (int attempt = 0; attempt < 30; ++attempt) {
          if (try_pull_frame()) {
            connected = true;
            encoding_ = "rgb8";
            RCLCPP_INFO(get_logger(), "Connected to RTSP fallback stream");
            break;
          }
          RCLCPP_WARN(get_logger(), "Waiting for RTSP fallback (attempt %d/30)...", attempt + 1);
          std::this_thread::sleep_for(1s);
        }
      }
    }

    if (!connected) {
      RCLCPP_FATAL(get_logger(), "Failed to connect to both SHM and RTSP streams");
      throw std::runtime_error("Camera stream connection failed");
    }

    // Load camera calibration via camera_info_manager
    std::string camera_info_url = "file://" + camera_info_file;
    camera_info_mgr_ = std::make_shared<camera_info_manager::CameraInfoManager>(
      this, "camera", camera_info_url);

    if (!camera_info_mgr_->isCalibrated()) {
      RCLCPP_WARN(get_logger(), "Camera not calibrated, using defaults");
    } else {
      RCLCPP_INFO(get_logger(), "Loaded camera calibration from %s", camera_info_file.c_str());
    }

    camera_info_ = camera_info_mgr_->getCameraInfo();
    camera_info_.width = width_;
    camera_info_.height = height_;
    camera_info_.header.frame_id = frame_id_;

    // Pre-allocate reusable Image message
    img_msg_.encoding = encoding_;
    img_msg_.width = width_;
    img_msg_.height = height_;
    img_msg_.step = width_ * 3;
    img_msg_.is_bigendian = false;
    img_msg_.data.resize(frame_size_);

    // Create publish timer
    auto period = std::chrono::duration<double>(1.0 / publish_rate);
    timer_ = create_wall_timer(
      std::chrono::duration_cast<std::chrono::nanoseconds>(period),
      std::bind(&CameraPublisherNode::publish_frame, this));

    RCLCPP_INFO(get_logger(), "Camera Publisher started at %.1f Hz (%s)", publish_rate, encoding_.c_str());
  }

  ~CameraPublisherNode()
  {
    cleanup_pipeline();
  }

private:
  bool try_pull_frame()
  {
    if (!appsink_) return false;
    GstSample *sample = gst_app_sink_try_pull_sample(GST_APP_SINK(appsink_), GST_SECOND);
    if (sample) {
      gst_sample_unref(sample);
      return true;
    }
    return false;
  }

  void cleanup_pipeline()
  {
    if (appsink_) {
      gst_object_unref(appsink_);
      appsink_ = nullptr;
    }
    if (pipeline_) {
      gst_element_set_state(pipeline_, GST_STATE_NULL);
      gst_object_unref(pipeline_);
      pipeline_ = nullptr;
    }
  }

  void publish_frame()
  {
    GstSample *sample = gst_app_sink_try_pull_sample(GST_APP_SINK(appsink_), 0);
    if (!sample) {
      ++read_fail_count_;
      if (read_fail_count_ % 60 == 0) {
        RCLCPP_WARN(get_logger(), "Failed to read frame (%d consecutive fails)", read_fail_count_);
      }
      return;
    }
    read_fail_count_ = 0;

    GstBuffer *buffer = gst_sample_get_buffer(sample);
    GstMapInfo map;
    if (!gst_buffer_map(buffer, &map, GST_MAP_READ)) {
      gst_sample_unref(sample);
      return;
    }

    // Stamp header
    auto stamp = now();
    img_msg_.header.stamp = stamp;
    img_msg_.header.frame_id = frame_id_;

    // Copy frame data directly into pre-allocated message buffer
    size_t copy_size = std::min(static_cast<size_t>(map.size), static_cast<size_t>(frame_size_));
    std::memcpy(img_msg_.data.data(), map.data, copy_size);

    gst_buffer_unmap(buffer, &map);
    gst_sample_unref(sample);

    // Sync camera_info timestamp
    camera_info_.header.stamp = stamp;

    image_pub_->publish(img_msg_);
    camera_info_pub_->publish(camera_info_);

    ++frame_count_;
    if (frame_count_ % 30 == 0) {
      RCLCPP_DEBUG(get_logger(), "Published %d frames", frame_count_);
    }
  }

  GstElement *pipeline_ = nullptr;
  GstElement *appsink_ = nullptr;
  std::string frame_id_;
  std::string rtsp_url_;
  std::string encoding_;
  int width_ = 0;
  int height_ = 0;
  int frame_size_ = 0;

  sensor_msgs::msg::Image img_msg_;
  sensor_msgs::msg::CameraInfo camera_info_;
  std::shared_ptr<camera_info_manager::CameraInfoManager> camera_info_mgr_;

  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr image_pub_;
  rclcpp::Publisher<sensor_msgs::msg::CameraInfo>::SharedPtr camera_info_pub_;
  rclcpp::TimerBase::SharedPtr timer_;

  int frame_count_ = 0;
  int read_fail_count_ = 0;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<CameraPublisherNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
